### DISCUSSION TITLE
CSR build, send it to a api, process the response and use it.

### DISCUSSION THREAD

**Discussion by lbtn** [Hakky54/ayza#124] [General]
Hello folks!

First of all, thanks for the great work!

I unfortunately have some really nasty scalability requirements on a project and even with a lot of reading I'm still pending to start a discussion. Hope it can help me and others =)

I made a mTLS communication with a server using Spring RestTemplate. It is working flawlessly. Great.
But now we have a scalability problem. We have multiple customers, each one receives a client_id and a temporary token from a service provider (by mail).
I'm building manually each .CSR file (with the mail information) and sending it, by postman to their API. They send me back a CRT response. I'm building it to a file and importing it to a keystore, using keytool. One by one.

So my first answer is related to this (I'm trying bouncy castle right now). Can I turn this process programmable?
Also, I need to store this CRT responses on a keystore? I can't simply store it in memory?

Thanks! Sorry if there is any basic knowledge missing, I'm trying really hard to understand it all.

**Comment by Hakky54** [comment]
Hello, 

I think it is possible to fully automate this, but I am not sure as I need some code examples or example repository to try it out myself.

Your flow looks like this:
- Build a CSR programatically -> https://stackoverflow.com/a/20550258/6777695
- Signing the CSR -> Send a POST request with the CSR in the body to their API with a http client in java. Get the response body as inputstream and transform it into a certificate object.
- Repeat and store into a Keystore -> You can collect all the certificates during runtime when you are finished you can write them to an in-memory KeyStore object and then finally write this object to a file on your file system. This file will than contain all of the certificates.
